src/coloring.jl

```diff
@@ -401,4 +413,159 @@ $TYPEDFIELDS
+
+        # Update the neighbors of i in the tree T
+        if !haskey(trees[index_tree], i)
+            trees[index_tree][i] = [j]
+        else
+            push!(trees[index_tree][i], j)
+        end
+
+        # Update the neighbors of j in the tree T
+        if !haskey(trees[index_tree], j)
+            trees[index_tree][j] = [i]
+        else
+            push!(trees[index_tree][j], i)
+        end
+    end
+
+    # degrees is a vector of integers that stores the degree of each vertex in a tree
+    degrees = Vector{Int}(undef, nvertices)
+
+    # list of vertices for each tree in the forest
+    vertices_by_tree = [collect(keys(trees[i])) for i in 1:ntrees]
+
+    # reverse breadth first (BFS) traversal order for each tree in the forest
+    reverse_bfs_orders = [Tuple{Int,Int}[] for i in 1:ntrees]
+
+    # nvmax is the number of vertices of the biggest tree in the forest
+    nvmax = mapreduce(length, max, vertices_by_tree; init=0)
+
+    # Create a queue with a fixed size nvmax
+    queue = Vector{Int}(undef, nvmax)
+
+    for k in 1:ntrees
+        tree = trees[k]
+
+        # Initialize the queue to store the leaves
+        queue_start = 1
+        queue_end = 0
+
+        # compute the degree of each vertex in the tree
+        for (vertex, neighbors) in tree
+            degree = length(neighbors)
+            degrees[vertex] = degree
+
+            # the vertex is a leaf
+            if degree == 1
+                queue_end += 1
+                queue[queue_end] = vertex
+            end
+        end
+
+        # continue until all leaves are treated
+        while queue_start <= queue_end
+            leaf = queue[queue_start]
+            queue_start += 1
+
+            # Mark the vertex as removed
+            degrees[leaf] = 0
+
+            for neighbor in tree[leaf]
+                if degrees[neighbor] != 0
+                    # (leaf, neighbor) represents the next edge to visit during decompression
+                    push!(reverse_bfs_orders[k], (leaf, neighbor))
+
+                    # reduce the degree of the neighbor
+                    degrees[neighbor] -= 1
+
+                    # check if the neighbor is now a leaf
+                    if degrees[neighbor] == 1
+                        queue_end += 1
+                        queue[queue_end] = neighbor
+                    end
+                end
+            end
+        end
+    end
+
+    return TreeSet(forest, vertices_by_tree, reverse_bfs_orders)
+end
+
+## Postprocessing, mirrors decompression code
+
+function postprocess!(
+    color::AbstractVector{<:Integer},
+    star_or_tree_set::Union{StarSet,TreeSet},
+    g::AdjacencyGraph,
+)
+    (; S) = g
+    # flag which colors are actually used during decompression
+    color_used = zeros(Bool, maximum(color))
+    # nonzero diagonal coefficients force the use of their respective color (there can be no neutral colors if the diagonal is fully nonzero)
+    for i in axes(S, 1)
```

Comment:
Don't we already have the indexes of the diagonal coefficients when we build our adjacency graph?

Reply:
we can retrieve this info from the graph instead of the matrix-like object, but given that we never store it explicitly I don't think we can go much faster than the way it's written here (which also mirrors decompression)